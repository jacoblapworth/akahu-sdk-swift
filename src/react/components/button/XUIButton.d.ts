import * as React from 'react';

import { XUIIconData } from '../icon/XUIIcon';
import {
  buttonTypes,
  sizeClassNames,
  textButtonVariants,
  widthClassNames,
} from './private/constants';

interface BaseProps {
  /**
   * Optional prop for users to modify the Button caret icon, if required for localisation.
   * Defaults to the caret icon, if no value is provided.
   */
  caretIcon?: XUIIconData;
  children?: React.ReactNode;
  className?: string;
  /**
   * Modifier for the width of the button.
   */
  fullWidth?: keyof typeof widthClassNames;
  /**
   * Use this to specify whether this button will have a dropdown caret.
   */
  hasCaret?: boolean;
  /**
   * The `href` attribute to use on the anchor element.
   *
   * Ignored unless `isLink` is `true`.
   */
  href?: string;
  /**
   * Determines if the button is disabled or not.
   */
  isDisabled?: boolean;
  /**
   * If true, sets appropriate `rel` values to prevent new page from having access to
   * `window.opener`. Should be used for links pointing at external sites.
   *
   * Ignored unless `isLink` is `true`.
   */
  isExternalLink?: boolean;
  /**
   * If this button is part of a parent button group.
   */
  isGrouped?: boolean;
  /**
   * Applies inverted styling.
   */
  isInverted?: boolean;
  /**
   * Whether or not to render this button using an anchor tag.
   */
  isLink?: boolean;
  /**
   * If true, shows a loader inside the button and also disables the button to prevent clicking. Can
   * be used in conjunction with `isDisabled` (which also provides a disabled class).
   */
  isLoading?: boolean;
  /**
   * Icon to appear to the left of the button content.
   */
  leftIcon?: React.ReactNode;
  /**
   * Accessibility label for the `XUILoader`. This is required if the `isLoading` prop is set to
   * `true`.
   */
  loadingAriaLabel?: string;
  /**
   * Use this to specify a min width on the button, when you want to swap to loading states.
   */
  minLoaderWidth?: boolean;
  /**
   * Bind a function to fire when the button is clicked.
   */
  onClick?: React.MouseEventHandler<HTMLElement>;
  /**
   * A keydown event handler for the button.
   */
  onKeyDown?: React.KeyboardEventHandler<HTMLElement>;
  qaHook?: string;
  /**
   * The `rel` attribute to use on the anchor element.
   *
   * Ignored unless `isLink` is `true`.
   */
  rel?: string;
  /**
   * When used with `isLoading` this allows the button to retain children width.
   */
  retainLayout?: boolean;
  /**
   * Icon to appear to the right of the button content.
   */
  rightIcon?: React.ReactNode;
  /**
   * Modifier for the size of the button. Buttons with `variant` set to `unstyled` will ignore the
   * `size` property.
   */
  size?: keyof typeof sizeClassNames;
  /**
   * The HTML `tabIndex` attribute value.
   */
  tabIndex?: number;
  /**
   * The `target` attribute to use on the anchor element.
   *
   * Ignored unless `isLink` is `true`.
   */
  target?: string;
  /**
   * The `title` attribute.
   */
  title?: string;
  /**
   * The type attribute of this button.
   *
   * Ignored unless `isLink` is `false`.
   */
  type?: keyof typeof buttonTypes;
  /**
   * Determines the styling variation to apply.
   */
  variant?: keyof typeof textButtonVariants;
}

type SpreadProps =
  | React.AnchorHTMLAttributes<HTMLAnchorElement>
  | React.ButtonHTMLAttributes<HTMLButtonElement>;

type Props = BaseProps & SpreadProps;

export default class XUIButton extends React.PureComponent<Props> {
  /**
   * Focus the button.
   */
  focus(): void;
  /**
   * Check if the button has focus.
   */
  hasFocus(): boolean;
  /**
   * Root node of XUICompleter to enable users to access as a ref.
   */
  rootNode: HTMLElement | null;
}
