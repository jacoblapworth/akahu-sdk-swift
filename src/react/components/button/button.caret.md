The trigger to open a [`<DropDownToggled>`](#dropdown) is often a Button. We provide a `<XUIButtonCaret />` component so you don't need to manually add a caret icon each time.

Check out the [Dropdown](#dropdown) examples to see Buttons being used as triggers.

```jsx
	<XUIButton>
		Dropdown <XUIButtonCaret />
	</XUIButton>
```

But you can just as easily drop in other icons too.

```jsx
	const XUIIcon = require('../icon/XUIIcon').default;
	const icon = require('@xero/xui-icon/icons/checkbox-check').default;
	<XUIButton>
		Checkbox <XUIIcon icon={icon} />
	</XUIButton>
```
