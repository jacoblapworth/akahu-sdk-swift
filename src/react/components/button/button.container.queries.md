For an example of how to switch between button types responsively, see [Container Queries](#container-queries).
