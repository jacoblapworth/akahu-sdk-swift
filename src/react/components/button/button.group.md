Buttons can be grouped together (e.g. if their actions are all related) by using a `<XUIButtonGroup>`.

To disable an entire button group, you must add the `isDisabled` prop to each button in the group, not to the `<XUIButtonGroup>` itself.

```
	<XUIButtonGroup qaHook="numbered-buttongroup">
		<XUIButton>One</XUIButton>
		<XUIButton>Two</XUIButton>
		<XUIButton>Three</XUIButton>
	</XUIButtonGroup>
```

To set the size of an entire button group, you can set the `size` prop on `<XUIButtonGroup>`.

```
	<XUIButtonGroup qaHook="numbered-buttongroup" size="small">
		<XUIButton>One</XUIButton>
		<XUIButton>Two</XUIButton>
		<XUIButton>Three</XUIButton>
	</XUIButtonGroup>
```
