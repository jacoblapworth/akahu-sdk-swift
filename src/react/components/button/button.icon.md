For buttons that only contain an icon, use `<XUIIconButton />`.

The `icon` and `ariaLabel` props are required when using this component.

Inverted style will be applied when `isInverted` prop is passed.

The `size` prop is used for the size classes of the iconButton, and the `iconSize` prop is used for the size classes of the icon.

```jsx harmony
import overflow from '@xero/xui-icon/icons/overflow';
import { XUIIconButton } from '@xero/xui/react/button';

import ExampleContainer from '../../docs/ExampleContainer';

const props = {
  ariaLabel: 'Overflow menu',
  icon: overflow
};

<div>
  <ExampleContainer className="xui-padding-xsmall">
    <XUIIconButton {...props} />
    <br />
    <XUIIconButton {...props} size="small" />
    <br />
    <XUIIconButton {...props} size="xsmall" />
  </ExampleContainer>

  <ExampleContainer className="xui-padding-xsmall" isInverted>
    <XUIIconButton {...props} isInverted />
    <br />
    <XUIIconButton {...props} isInverted size="small" />
    <br />
    <XUIIconButton {...props} isInverted size="xsmall" />
  </ExampleContainer>
</div>;
```
