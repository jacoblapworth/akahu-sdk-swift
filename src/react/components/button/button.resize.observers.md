For an example of how to switch between button types responsively, see [Resize Observers](#resize-observers).
