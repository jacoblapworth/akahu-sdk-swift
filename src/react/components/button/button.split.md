A split button is used the present a primary action coupled with a dropdown trigger for secondary actions.
The `<XUISplitButtonGroup>` component will inherit the `isDisabled` and the `variant` props from the parent component down to the button children.

A split button can only be completely disabled - you cannot disable only one part of the button

```
// try setting `isDisabled={true}`, or change the variant, and see how both buttons are disabled

<XUISplitButtonGroup variant="primary" isDisabled={false} >
	<XUIButton>Split Button</XUIButton>
	<XUISplitButton />
</XUISplitButtonGroup>
```
