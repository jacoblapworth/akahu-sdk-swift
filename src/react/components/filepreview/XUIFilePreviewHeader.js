import React from 'react';
import PropTypes from 'prop-types';
import cn from 'classnames';
import { ns } from '../helpers/xuiClassNamespace';
import shouldRender from '../helpers/shouldRender';
import useContainerQuery from '../../helpers/useContainerQuery';

const baseClass = `${ns}-filepreviewheader`;

const XUIFilePreviewHeader = ({
  actions,
  children,
  className,
  headerTag,
  headingLevel,
  navigationButton,
  qaHook,
  secondary,
  title,
  ...spreadProps
}) => {
  const { isWidthAboveBreakpoint, observedElementRef } = useContainerQuery();

  const classes = cn(
    className,
    baseClass,
    isWidthAboveBreakpoint('small') && `${baseClass}-medium-up`,
  );
  const HeaderElem = headerTag;
  const HeadingTag = `h${headingLevel}`;

  const controlContent = shouldRender(navigationButton) && (
    <div className={`${baseClass}--controlcontent`}>{navigationButton}</div>
  );

  const titleTags = (
    <div className={`${baseClass}--titlewrapper`}>
      <HeadingTag
        className={`${baseClass}--title`}
        data-automationid={qaHook && `${qaHook}--title`}
      >
        {title}
      </HeadingTag>
      {shouldRender(secondary) && (
        <div
          className={`${baseClass}--secondarytitle`}
          data-automationid={qaHook && `${qaHook}--secondarytitle`}
        >
          {secondary}
        </div>
      )}
    </div>
  );
  const leftContent = (
    <div className={`${baseClass}--leftcontent`}>
      {controlContent}
      {titleTags}
    </div>
  );
  const rightContent = shouldRender(actions) && (
    <div className={`${baseClass}--rightcontent`}>
      {actions && <div className={`${baseClass}--actions`}>{actions}</div>}
    </div>
  );
  return (
    <HeaderElem
      {...spreadProps}
      className={classes}
      data-automationid={qaHook}
      ref={observedElementRef}
    >
      {leftContent}
      {children}
      {rightContent}
    </HeaderElem>
  );
};
export default XUIFilePreviewHeader;

XUIFilePreviewHeader.propTypes = {
  /**
   * Components or html to be right-aligned in the filepreviewheader
   */
  actions: PropTypes.node,
  /**
   * Content to go in the header. It's unlikely that you will need to use this.
   */
  children: PropTypes.node,
  /**
   * Classes to be applied to the filepreviewheader element
   */
  className: PropTypes.string,
  /**
   * Tag type for the filepreviewheader element. Defaults to 'header'
   */
  headerTag: PropTypes.string,
  /**
   * The level of heading that should be assigned to the title. Defaults to 2.
   */
  headingLevel: PropTypes.number,
  /**
   * Navigation button to appear to the left of the heading. Used to close/minimise the preview.
   */
  navigationButton: PropTypes.node,
  qaHook: PropTypes.string,
  /**
   * Secondary title
   */
  secondary: PropTypes.node,
  /**
   * Title text or node. Required.
   */
  title: PropTypes.node.isRequired,
};

XUIFilePreviewHeader.defaultProps = {
  headerTag: 'header',
  headingLevel: 2,
};
