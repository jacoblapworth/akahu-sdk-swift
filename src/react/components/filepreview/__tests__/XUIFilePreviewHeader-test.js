import React from 'react';
import { axe, toHaveNoViolations } from 'jest-axe';
import { render } from '@testing-library/react';

import cross from '@xero/xui-icon/icons/cross';
import importIcon from '@xero/xui-icon/icons/import';
import XUIFilePreviewHeader from '../XUIFilePreviewHeader';
import { XUIButton, XUIIconButton } from '../../../button';
import { XUIIcon } from '../../../icon';

expect.extend(toHaveNoViolations);

const createComponent = props => (
  <XUIFilePreviewHeader title="test title" {...props}></XUIFilePreviewHeader>
);

const navigationButton = <XUIIconButton ariaLabel="close" icon={cross} onClick={() => {}} />;
const downloadAction = (
  <XUIButton size="small" variant="borderless-main">
    <XUIIcon icon={importIcon} />
    Download
  </XUIButton>
);

describe('<XUI XUIFilePreviewHeader/>', () => {
  test('renders basic example correctly', () => {
    const { asFragment } = render(createComponent());
    expect(asFragment()).toMatchSnapshot();
  });

  test('renders a passed qaHook as an automationId', () => {
    const { asFragment } = render(createComponent({ qaHook: 'test-filepreviewheader' }));
    expect(asFragment()).toMatchSnapshot();
  });

  test('renders any additional classes provided through the className prop', () => {
    const { asFragment } = render(createComponent({ className: 'filepreviewheader-class' }));
    expect(asFragment()).toMatchSnapshot();
  });

  test('renders alternate wrapping element, if provided', () => {
    const { asFragment } = render(createComponent({ headerTag: 'article' }));
    expect(asFragment()).toMatchSnapshot();
  });

  test('renders with all content supplied', () => {
    const { asFragment } = render(
      createComponent({
        actions: downloadAction,
        children: <a>link</a>,
        headingLevel: 3,
        navigationButton,
        secondary: 'Secondary title',
        title: 'Required title',
      }),
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test('should pass accessibility testing', async () => {
    const { container } = render(createComponent());
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });
});
