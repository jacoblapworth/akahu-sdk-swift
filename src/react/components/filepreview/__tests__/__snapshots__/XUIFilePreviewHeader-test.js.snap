// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<XUI XUIFilePreviewHeader/> renders a passed qaHook as an automationId 1`] = `
<DocumentFragment>
  <header
    class="xui-filepreviewheader xui-filepreviewheader-medium-up"
    data-automationid="test-filepreviewheader"
  >
    <div
      class="xui-filepreviewheader--leftcontent"
    >
      <div
        class="xui-filepreviewheader--titlewrapper"
      >
        <h2
          class="xui-filepreviewheader--title"
          data-automationid="test-filepreviewheader--title"
        >
          test title
        </h2>
      </div>
    </div>
  </header>
</DocumentFragment>
`;

exports[`<XUI XUIFilePreviewHeader/> renders alternate wrapping element, if provided 1`] = `
<DocumentFragment>
  <article
    class="xui-filepreviewheader xui-filepreviewheader-medium-up"
  >
    <div
      class="xui-filepreviewheader--leftcontent"
    >
      <div
        class="xui-filepreviewheader--titlewrapper"
      >
        <h2
          class="xui-filepreviewheader--title"
        >
          test title
        </h2>
      </div>
    </div>
  </article>
</DocumentFragment>
`;

exports[`<XUI XUIFilePreviewHeader/> renders any additional classes provided through the className prop 1`] = `
<DocumentFragment>
  <header
    class="filepreviewheader-class xui-filepreviewheader xui-filepreviewheader-medium-up"
  >
    <div
      class="xui-filepreviewheader--leftcontent"
    >
      <div
        class="xui-filepreviewheader--titlewrapper"
      >
        <h2
          class="xui-filepreviewheader--title"
        >
          test title
        </h2>
      </div>
    </div>
  </header>
</DocumentFragment>
`;

exports[`<XUI XUIFilePreviewHeader/> renders basic example correctly 1`] = `
<DocumentFragment>
  <header
    class="xui-filepreviewheader xui-filepreviewheader-medium-up"
  >
    <div
      class="xui-filepreviewheader--leftcontent"
    >
      <div
        class="xui-filepreviewheader--titlewrapper"
      >
        <h2
          class="xui-filepreviewheader--title"
        >
          test title
        </h2>
      </div>
    </div>
  </header>
</DocumentFragment>
`;

exports[`<XUI XUIFilePreviewHeader/> renders with all content supplied 1`] = `
<DocumentFragment>
  <header
    class="xui-filepreviewheader xui-filepreviewheader-medium-up"
  >
    <div
      class="xui-filepreviewheader--leftcontent"
    >
      <div
        class="xui-filepreviewheader--controlcontent"
      >
        <button
          aria-label="close"
          class="xui-button xui-button-icon-medium xui-button-icon"
          tabindex="0"
          type="button"
        >
          <svg
            class="xui-icon"
            focusable="false"
            height="12"
            viewBox="0 0 12 12"
            width="12"
          >
            <path
              d="M6 4.5L10.5 0 12 1.5 7.5 6l4.5 4.5-1.5 1.5L6 7.5 1.5 12 0 10.5 4.5 6 0 1.5 1.5 0 6 4.5z"
              role="presentation"
            />
          </svg>
          <div
            class="xui-touchtarget"
          />
        </button>
      </div>
      <div
        class="xui-filepreviewheader--titlewrapper"
      >
        <h3
          class="xui-filepreviewheader--title"
        >
          Required title
        </h3>
        <div
          class="xui-filepreviewheader--secondarytitle"
        >
          Secondary title
        </div>
      </div>
    </div>
    <a>
      link
    </a>
    <div
      class="xui-filepreviewheader--rightcontent"
    >
      <div
        class="xui-filepreviewheader--actions"
      >
        <button
          class="xui-button xui-button-borderless-main xui-button-small"
          tabindex="0"
          type="button"
        >
          <svg
            class="xui-icon"
            focusable="false"
            height="12"
            viewBox="0 0 10 12"
            width="10"
          >
            <path
              d="M4 5V0h2v5h2L5 8 2 5h2zm-4 6h10v1H0v-1z"
              role="presentation"
            />
          </svg>
          Download
        </button>
      </div>
    </div>
  </header>
</DocumentFragment>
`;
